### DISCUSSION TITLE
CHM edge buffer

### DISCUSSION THREAD

**Discussion by djes** [r-lidar/lidR#647] [Ideas]
Hi,

I have a request. When using rasterize_canopy(), the resulting CHM has no buffer. It's OK if one use it as it is. But if you want to modify the rasters, for example with focal(), maybe it would be useful to have a buffered CHM...?

Here's two pictures showing the artefact on using focal on non-buffered CHM.
![image](https://user-images.githubusercontent.com/12065874/205026939-db631429-65f9-4a65-8458-1678aa421fc2.png)
![image](https://user-images.githubusercontent.com/12065874/205026230-c0baf73a-34df-40c3-89de-ba66951c7b00.png)
 
I've used this basic code to use focal().
```r
f = list.files("7_chm/", pattern = "*.tif")

print(paste0("Nb: ", length(f)))

for( i in 1:length(f)){
	ker = matrix(1, 5, 5)
  	r <- rast(paste0("7_chm/", f[i]))
	nchm <- focal(r, w = ker, fun = median, na.rm = TRUE)
	writeRaster(nchm, paste0("7_chm/focal5x5/", f[i]))
	print(i)
	flush.console()
}
```

**Comment by Jean-Romain** [comment]
I do not understand the problem. The output CHM is expected to be perfectly continuous. What's wrong with it?

The problem is that you are looping over the files independently instead of working with one large composite raster. You should not loop over the tif file for the same reason you should not loop over las files.

**Reply by djes** [reply]
Thank you Jean-Romain. It is not really a problem, just a feature request. Working with a large composite raster is not always possible, due to memory concerns. If it was possible to handle buffered tiles, there wouldn't be any edge problem. I just discovered that my segmentation was faulty due to this.

**Reply by Jean-Romain** [reply]
Something like that should do the job

```r
opt_chunk_buffer(ctg) = 50
opt_chunk_size(ctg) = 0
opt_output_files(ctg) = "..."
chunks = engine_chunks(ctg)
for (chunk in chunks)
{
    las = readLAS(chunk)
    chm = rasterize_canopy(las, ...)
    writeRaster(chm, chunk@save)
}
```

**Reply by djes** [reply]
Perfect, thank you for this example.
